File: src/metrics.py
"""
Degradation metrics tracker for LLM sessions.

Monitors signal frequency and patterns to detect
when a model's context is becoming unreliable.
"""

from collections import defaultdict
from datetime import datetime


class SessionMetrics:
    def __init__(self):
        self.signal_counts = defaultdict(int)
        self.signal_history = []  # (timestamp, signal)
        self.session_start = datetime.now()

        # Thresholds that trigger degradation warning
        self.degradation_thresholds = {
            "@@UNKNOWN": 3,   # 3 unknowns = likely degraded
            "@@ESCALATE": 5,  # 5 escalations = model struggling
            "total": 15,      # 15 signals total = long session
        }

    def record(self, signal: str):
        """Record a signal occurrence."""
        self.signal_counts[signal] += 1
        self.signal_history.append((datetime.now(), signal))

    def check_degradation(self) -> list[str]:
        """
        Check if degradation thresholds are exceeded.
        Returns list of warning messages, empty if healthy.
        """
        warnings = []

        for signal, threshold in self.degradation_thresholds.items():
            if signal == "total":
                total = sum(self.signal_counts.values())
                if total >= threshold:
                    warnings.append(
                        f"High signal volume: {total} total signals this session"
                    )
            elif self.signal_counts[signal] >= threshold:
                warnings.append(
                    f"Threshold exceeded: {signal} occurred "
                    f"{self.signal_counts[signal]} times"
                )

        return warnings

    def consecutive_unknowns(self) -> int:
        """Count consecutive @@UNKNOWN signals at end of history."""
        count = 0
        for _, signal in reversed(self.signal_history):
            if signal == "@@UNKNOWN":
                count += 1
            else:
                break
        return count

    def summary(self) -> str:
        """Return a human-readable session summary."""
        duration = datetime.now() - self.session_start
        minutes = int(duration.total_seconds() / 60)
        lines = [
            f"Session duration: {minutes} minutes",
            f"Total signals: {sum(self.signal_counts.values())}",
        ]
        for signal, count in self.signal_counts.items():
            lines.append(f"  {signal}: {count}")
        return '\n'.join(lines)

File: src/metrics.js
/**
 * Session metrics tracker. JS port of metrics.py.
 */

class SessionMetrics {
    constructor() {
        this.signalCounts = {};
        this.signalHistory = []; // [{timestamp, signal}]
        this.sessionStart = new Date();
        this.thresholds = {
            '@@UNKNOWN': 3,
            '@@ESCALATE': 5,
            total: 15,
        };
    }

    record(signal) {
        this.signalCounts[signal] = (this.signalCounts[signal] || 0) + 1;
        this.signalHistory.push({ timestamp: new Date(), signal });
    }

    checkDegradation() {
        const warnings = [];
        const total = Object.values(this.signalCounts).reduce((a, b) => a + b, 0);

        if (total >= this.thresholds.total) {
            warnings.push(`High signal volume: ${total} total signals`);
        }

        for (const [signal, threshold] of Object.entries(this.thresholds)) {
            if (signal === 'total') continue;
            if ((this.signalCounts[signal] || 0) >= threshold) {
                warnings.push(
                    `${signal} threshold exceeded: ${this.signalCounts[signal]} times`
                );
            }
        }

        return warnings;
    }

    consecutiveUnknowns() {
        let count = 0;
        for (let i = this.signalHistory.length - 1; i >= 0; i--) {
            if (this.signalHistory[i].signal === '@@UNKNOWN') count++;
            else break;
        }
        return count;
    }

    summary(compact = false) {
        const duration = Math.floor((new Date() - this.sessionStart) / 60000);
        if (compact) {
            const total = Object.values(this.signalCounts).reduce((a, b) => a + b, 0);
            return `${duration}min, ${total} signals`;
        }
        const lines = [`Duration: ${duration} minutes`];
        for (const [signal, count] of Object.entries(this.signalCounts)) {
            lines.push(`  ${signal}: ${count}`);
        }
        return lines.join('\n');
    }
}

module.exports = { SessionMetrics };